### PR TITLE
updated repo with mobilenet_v4

### LINKED ISSUE #4289: Introduce MobileNetV4 for Improved Performance
MobileNetV4 (https://arxiv.org/pdf/2404.10518) has been introduced as a faster and more accurate alternative to MobileNetV3. The goal of this issue is to integrate MobileNetV4 into our repository to replace the older MobileNetV3 model. Since MobileNetV4 is designed to be both efficient and effective, we will test its performance across diverse datasets, with a particular focus on tiny datasets starting from as few as 6–10 images.

Tasks:
 1. Integrate MobileNetV4:
     • Implement MobileNetV4 starting with the Timm (https://github.com/huggingface/pytorch-image-models) implementation. We have already integration with timm backbones: (https://github.com/openvinotoolkit/training_extensions/blob/develop/src/otx/algo/classification/backbones/timm.py) : 
     • Ensure the implementation follows the existing code structure. 
    - It is also possible to implement it from scratch if any issues with timm using different sources. 

 2. Benchmark Performance:
     • Compare accuracy, speed, and efficiency against MobileNetV3.
     • Test on a variety of datasets, emphasizing tiny datasets (6–10 images) to evaluate performance in low-data scenarios.
 3. Optimize for Small Datasets:
     • Explore training strategies to improve performance with limited data (up to 100) (e.g., data augmentation, transfer learning, regularization). Meanwhile preserve competitive accuracy on larger datasets (1000+) 
 4. Validate and Document Findings:
     • Provide performance reports with quantitative comparisons.
     • Provide config (recipe) for e2e solution. (from torch fine-tuning to OpenVINO IR export/testing)

### PR DESCRIPTION
**Pull request by GauravSRC**
Summary

Resolves https://github.com/openvinotoolkit/training_extensions/issues/4289
Introduces MobileNetV4 integration into OTX framework for enhanced multi-class classification

Key Contributions:
- Integrated MobileNetV4 using TIMM backbone with flexible implementation
- Developed custom `MobileNetV4MulticlassCls` supporting multiple variants
- Implemented advanced performance optimization strategies

Detailed Implementation
I integrated MobileNetV4 into the OTX framework, creating a flexible multi-class classification model. By utilizing the TIMM backbone and developing a custom MobileNetV4MulticlassCls class, I implemented a robust solution that supports multiple model variants (small, medium, large) with advanced features like label smoothing, dropout regularization, and TorchScript compilation support.
Through comprehensive benchmarking across tiny (6-10 images), small (500 images), and full CIFAR-10 datasets, I demonstrated MobileNetV4's performance improvements compared to MobileNetV3. The results revealed incremental accuracy gains: on the tiny dataset, MobileNetV4 achieved 39.06% accuracy versus MobileNetV3's 35.89%, and on the small dataset, the performance increased to 67.81% versus 59.52%. I implemented optimized training strategies including mixed-precision training, adaptive early stopping, warmup learning rate scheduling, and sophisticated data augmentation techniques to enhance performance, particularly in low-data scenarios.
Also i have mentioned the key observations, the challenges I am facing and the modifications i have started with in the last comment so Looking forward to your insights and recommendations.

How to Test
Validate implementation by:
Training on CIFAR-10 dataset with different data scales
Comparing performance metrics between MobileNetV4 and MobileNetV3
Exporting and testing ONNX and OpenVINO IR model formats

Checklist

 Implemented MobileNetV4 integration
 Added performance benchmarks
 Developed model export capabilities

License

 Code submitted under Apache License 2.0

### CODE REVIEW

**Comment by sovrasov** [comment]
@GauravSRC please run `tox -vv -e pre-commit` to fix basic issues highlighted by linters

File: src/otx/algo/classification/multiclass_models/mobilenet_v4.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        optimizer (OptimizerCallable, optional): Custom optimizer configuration
+        scheduler (LRSchedulerCallable, optional): Custom learning rate scheduler
+        metric (MetricCallable, optional): Custom evaluation metric
+        torch_compile (bool, optional): Enable TorchScript compilation
+    """
+    def __init__(
+        self,
+        label_info: LabelInfoTypes,
+        data_input_params: DataInputParams,
+        model_name: str = 'mobilenetv4_conv_medium',
```
Comment (sovrasov) [inline src/otx/algo/classification/multiclass_models/mobilenet_v4.py]: Is it possible to remove this file and use `src/otx/algo/classification/multiclass_models/timm_model.py` with a corresponding model name provided via recipe?

File: src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  class_path: otx.algo.classification.multiclass_models.timm_model.TimmModelMulticlassCls
+  init_args:
+    model_name: mobilenetv4_conv_medium  # Using the medium variant from Timm
+    label_info: 1000  # Default number of classes, can be adjusted
+    optimizer:
+      class_path: torch.optim.AdamW  # Modern optimizer with weight decay
+      init_args:
+        lr: 0.0005  # Slightly lower learning rate for MobileNetV4
+        weight_decay: 0.01  # Regularization to prevent overfitting
+        betas: [0.9, 0.999]  # Standard AdamW parameters
```
Comment (sovrasov) [inline src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml]: If some params are already default ones, we might just omit them in the config

File: src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml [hunk 3]
```diff
@@ -0,0 +1,63 @@
+model:
+  class_path: otx.algo.classification.multiclass_models.timm_model.TimmModelMulticlassCls
+  init_args:
+    model_name: mobilenetv4_conv_medium  # Using the medium variant from Timm
+    label_info: 1000  # Default number of classes, can be adjusted
```
Comment (sovrasov) [inline src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml]: Please remove unnecessary comments

File: src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml [hunk 4]
```diff
@@ -0,0 +1,63 @@
+model:
+  class_path: otx.algo.classification.multiclass_models.timm_model.TimmModelMulticlassCls
+  init_args:
+    model_name: mobilenetv4_conv_medium  # Using the medium variant from Timm
```
Comment (sovrasov) [inline src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml]: How are the flops and #params of the medium MNv4 model compared vs MNv3_large? Are these models in the same compute budget?

File: src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    - class_path: otx.algo.callbacks.adaptive_early_stopping.EarlyStoppingWithWarmup
+      init_args:
+        patience: 5
+        min_delta: 0.001  # Small improvement threshold
+    - class_path: lightning.pytorch.callbacks.ModelCheckpoint
+      init_args:
+        monitor: val/accuracy
+        mode: max
+        save_top_k: 3
+        filename: 'mobilenetv4-{epoch:02d}-{val/accuracy:.2f}'
```
Comment (sovrasov) [inline src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml]: Why does checkpoint name need to be modified?

File: src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+# Tiny dataset specific augmentations
+augmentations:
```
Comment (sovrasov) [inline src/otx/recipe/classification/multi_class_cls/mobilenet_v4.yaml]: Is there any specific reason not to use common classification data config?

**Review by sovrasov — COMMENTED** [review]
@GauravSRC thanks for the PR!
Since it's a new model for us, it'd be great to have some comparison vs MNv3_large. I see some results in the PR description, but they don't seem clear to me. 
You can refer to the tables in https://github.com/openvinotoolkit/training_extensions/issues/4286 to get some inspiration on how to provide model comparison results

**Comment by GauravSRC** [comment]
| Dataset      | Model        | Size  | Test Accuracy | Epochs | Best Val Acc | Training Time (s) |
|--------------|--------------|-------|----------------|--------|---------------|--------------------|
| CIFAR10      | MobileNetV4  | tiny  | 47.44%         | 34     | 65.00%        | 28.55              |
| CIFAR10      | MobileNetV4  | small | 71.83%         | 31     | 80.00%        | 86.48              |
| CIFAR10      | MobileNetV3  | tiny  | 45.79%         | 19     | 65.00%        | 12.04              |
| CIFAR10      | MobileNetV3  | small | 64.15%         | 16     | 75.00%        | 41.87              |
| FashionMNIST | MobileNetV4  | tiny  | 65.21%         | 30     | 75.00%        | 21.13              |
| FashionMNIST | MobileNetV4  | small | 82.34%         | 23     | 88.00%        | 69.60              |
| FashionMNIST | MobileNetV3  | tiny  | 60.94%         | 17     | 60.00%        | 11.58              |
| FashionMNIST | MobileNetV3  | small | 78.20%         | 17     | 87.00%        | 49.45              |

Enhanced Training Comparison on CIFAR10 (small):

| Model         | Regular Accuracy | Enhanced Accuracy | Improvement |
|---------------|------------------|-------------------|-------------|
| MobileNetV4   | 72.81%           | 73.52%            | +0.71%      |
| MobileNetV3   | 66.37%           | 73.07%            | +6.70%      |

My code establishes a comprehensive experimental pipeline that prepares CIFAR10 and FashionMNIST datasets (with varying sizes), creates classifiers based on MobileNetV4 and MobileNetV3, and compares their performance using metrics such as test accuracy, epoch count, and training time. The results reveal that MobileNetV4 generally achieves higher accuracy on both datasets, especially with larger (small) subsets. Additionally, the enhanced training strategies—with improved data augmentation and fine-tuning—provide a noticeable boost for MobileNetV3, highlighting an interesting trade-off between computational efficiency (MACs) and performance.  
I am also providing the link(https://colab.research.google.com/drive/1JaP84OMjIjONlqgLgA-L0dGQkgshMnUF?usp=sharing) of my colab notebook for reference.
I am using MobileNetV4 (conv_small)[parameters=2,505,834] and MobileNetV3 Small [parameters=2,842,666] which have comparable flops and parameters and  these models have nearly same compute budget.

File: src/otx/algo/classification/multiclass_models/mobilenet_v4.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+"""MobileNetV4 Classification Model for OTX Framework
```
Comment (sovrasov) [inline src/otx/algo/classification/multiclass_models/mobilenet_v4.py]: Do we still need this file? It looks identical to `src/otx/algo/classification/multiclass_models/timm_model.py`

Comment (GauravSRC) [reply src/otx/algo/classification/multiclass_models/mobilenet_v4.py]: Yes, you're right - these files are nearly identical. I created the file intending to make a specialized MobileNetV4 wrapper but ultimately implemented it almost exactly like the general TIMM wrapper.
The main differences are just in the docstrings, where I added MobileNetV4-specific documentation. Looking at it now, I realize we don't need two nearly identical implementations. We can simply use the existing TIMM wrapper for MobileNetV4 models as well, since the core functionality is the same. So if you will say so, I will remove this file and then commit again with only the recipe.

Comment (sovrasov) [reply src/otx/algo/classification/multiclass_models/mobilenet_v4.py]: Please, remove the extra file

**Review by sovrasov — APPROVED** [review]

**Comment by rajeshgangireddy** [comment]
Hi @GauravSRC ,
Thanks for the contribution. 
Looks good to me. 

I have two questions : 
1) How did you arrive at the learning rate value in the recipe? Have you checked if the hparams suggested in timm for mobilenetv4 (https://gist.github.com/rwightman/f6705cb65c03daeebca8aa129b1b94ad) are usable/better here ?

 
2) Have you compared if any there are improvements in any other datasets other than CIFAR, FashionMNIST ? Could you please run the comparison with two non-toy datasets. Any two datasets shown  [here](https://github.com/open-edge-platform/training_extensions/issues/4286#issuecomment-2740657510) would be good. 

Thanks

**Review by eugene123tw — APPROVED** [review]

**Comment by GauravSRC** [comment]
> Hi @GauravSRC , Thanks for the contribution. Looks good to me.
> 
> I have two questions :
> 
> 1. How did you arrive at the learning rate value in the recipe? Have you checked if the hparams suggested in timm for mobilenetv4 (https://gist.github.com/rwightman/f6705cb65c03daeebca8aa129b1b94ad) are usable/better here ?
> 2. Have you compared if any there are improvements in any other datasets other than CIFAR, FashionMNIST ? Could you please run the comparison with two non-toy datasets. Any two datasets shown  [here](https://github.com/open-edge-platform/training_extensions/issues/4286#issuecomment-2740657510) would be good.
> 
> Thanks

Hi @rajeshgangireddy,
Apologies for the delayed response — I was caught up with quizzes and viva exams.

Regarding the learning rate mentioned in my recipe: when I executed the code, I noticed that different learning rates were being applied due to the scheduler. I selected the one that was approximately consistent and also matched the learning rate used in the MobileNetV3-Large recipe, so I decided to keep it the same.

Additionally, I compared the models on two different datasets. The comparison table is shared below:

| Dataset       | Size  | Model       | Epochs Trained | Best Val Acc (%) | Test Acc (%) | Training Time (s) |
|---------------|-------|-------------|----------------|------------------|--------------|-------------------|
| Flowers102    | Tiny  | MobileNetV4 | 12             | 58.82            | 53.92        | 118.37            |
| Flowers102    | Tiny  | MobileNetV3 | 11             | 57.35            | 56.37        | 98.68             |
| Flowers102    | Small | MobileNetV4 | 29             | 70.59            | 68.14        | 265.01            |
| Flowers102    | Small | MobileNetV3 | 16             | 64.71            | 58.33        | 150.51            |
| OxfordIITPet  | Tiny  | MobileNetV3 | 14             | 62.00            | 68.00        | 85.00             |
| OxfordIITPet  | Tiny  | MobileNetV4 | 18             | 56.00            | 56.00        | 108.00            |
| OxfordIITPet  | Full  | MobileNetV3 | 19             | 85.60            | 80.43        | 649.00            |
| OxfordIITPet  | Full  | MobileNetV4 | 20             | 64.27            | 67.43        | 721.00            |
